aep/general/0134/aep.md.j2

```diff
@@ -0,0 +1,321 @@
+- If partial resource update is supported, a field mask **must** be included.
+  It **must** be of type `google.protobuf.FieldMask`, and it **must** be called
+  `update_mask`.
+  - The fields used in the field mask correspond to the resource being updated
+    (not the request message).
+  - The field **may** be required or optional. If it is required, it **must**
+    include the corresponding annotation. If optional, the service **must**
+    treat an omitted field mask as an implied field mask equivalent to all
+    fields that are populated (have a non-empty value).
+  - Update masks **must** support a special value `*`, meaning full replacement
+    (the equivalent of `PUT`).
+
+{% tab oas %}
+
+**Note:** OAS example not yet written.
+
+{% endtabs %}
+
+### Side effects
+
+In general, update methods are intended to update the data within the resource.
+Update methods **should not** trigger other side effects. Instead, side effects
+**should** be triggered by custom methods.
+
+In particular, this entails that [state fields][] **must not** be directly
+writable in update methods.
+
+### PATCH and PUT
+
+**TL;DR:** AEP-compliant APIs generally use the `PATCH` HTTP verb only, and do
+not support `PUT` requests.
+
+We standardize on `PATCH` because many organizations update stable APIs in
+place with backwards-compatible improvements. It is often necessary to add a
+new field to an existing resource, but this becomes a breaking change when
+using `PUT`.
+
+To illustrate this, consider a `PUT` request to a `Book` resource:
+
+    PUT /v1/publishers/123/books/456
+
+    {"title": "Mary Poppins", "author": "P.L. Travers"}
+
+Next consider that the resource is later augmented with a new field (here we
+add `rating`, and use a protobuf example without loss of generality):
+
+```proto
+message Book {
+  string title = 1;
+  string author = 2;
+
+  // Subsequently added to v1 in place...
+  int32 rating = 3;
+}
+```
+
+If a rating were set on a book and the existing `PUT` request were executed, it
+would wipe out the book's rating. In essence, a `PUT` request unintentionally
+would wipe out data because the previous version did not know about it.
+
+### Long-running update
+
+Some resources take longer to update a resource than is reasonable for a
+regular API request. In this situation, the API **should** use a long-running
+operation (AIP-151) instead:
+
+- The response type **must** be set to the resource (what the return type would
+  be if the RPC was not long-running).
+
+{% tab proto %}
+
+```proto
+rpc UpdateBook(UpdateBookRequest) returns (aep.api.Operation) {
+  option (google.api.http) = {
+    patch: "/v1/{book.name=publishers/*/books/*}"
+  };
+  option (aep.api.operation_info) = {
+    response_type: "Book"
+    metadata_type: "OperationMetadata"
+  };
+}
+```
+
+- Both the `response_type` and `metadata_type` fields **must** be specified.
+
+{% tab oas %}
+
+**Note:** OAS example not yet written.
+
+{% endtabs %}
+
+### Create or update
```

Comment:
this is a very ugly pattern - we should remove it in favor of PUT / Apply. The logic to implement this on the server gets pretty complicated too (since you have to consider errors that are encountered with create as well as update).

I'd prefer to just remove this guidance now if we agree, and I'll come back with the PUT AEP. But if we don't agree I can just file an issue.

Reply:
Is PUT a true replacement for this pattern?  Create-or-update doesn't erase existing fields like PUT does.

In the interest of time/deadlines I'd suggest a ticket to revisit this.

Reply:
Created https://github.com/aep-dev/aep.dev/issues/123.